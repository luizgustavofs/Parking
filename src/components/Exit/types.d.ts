export type ExitState =
  | 'default'
  | 'error'
  | 'payment'
  | 'exit'
  | 'loading'
  | 'success';

export type ExitProps = {
  plate: string;
  setPlate: (newPlate: string) => void;
};
