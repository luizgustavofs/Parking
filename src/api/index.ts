import axios from 'axios';

export const api = axios.create({
  baseURL: 'https://parking-lot-to-pfz.herokuapp.com',
  timeout: 5000,
  headers: {},
});
