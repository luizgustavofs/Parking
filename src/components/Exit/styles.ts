import styled from 'styled-components';

export const Container = styled.div`
  flex-direction: column;
  min-height: 300px;
  padding: 16px;
  width: 100%;

  background: #fff;

  @media (min-width: 1024px) {
    width: 400px;
  }
`;

export const ConfirmScreen = styled.div`
  width: 100%;
  height: 100%;
  right: 0;
  top: 0;
  background-color: #fff;
  z-index: 1;
  align-items: center;
  justify-content: center;
  flex-direction: column;

  img {
    width: 130px;
    height: 130px;
  }
`;

export const LoadingScreen = styled.div`
  width: 100%;
  height: 100%;
  right: 0;
  top: 0;
  background-color: #fff;
  z-index: 1;
  align-items: center;
  justify-content: center;
  flex-direction: column;

  @media (min-width: 1024px) {
  }
`;

export const DivModal = styled.div`
  width: 100vw;
  height: 100vh;
  top: 0;
  left: 0;
  right: 0;
  background: #0005;
  position: absolute;
  z-index: 1;
  justify-content: center;
  align-items: center;

  > div {
    padding: 16px;
    width: 90vw;
    min-height: 90vw;
    background: white;
    flex-direction: column;
    justify-content: center;
    align-items: center;

    h3:first-child {
      width: 170px;
      text-align: center;
      margin-bottom: 1px;
    }

    h1 {
      margin-top: 1px;
      color: ${({ theme }) => theme.colors.primary};
    }

    @media (min-width: 1024px) {
      width: 30vw;
      min-height: 30vw;
    }
  }
`;
