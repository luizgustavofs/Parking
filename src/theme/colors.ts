export const colors = {
  default: '#000000',
  primary: '#4DD0E1',
  secondary: '#F2F2F2',
  tertiary: '#28DD91',
  quaternary: '#A769B2',
  error: '#FF0068',
  baseColor: '#FFFBE6',
  neutral: {
    _00: '#ffffff',
    _70: '#DADADA',
    _80: '#9B9B9B',
  },
};
