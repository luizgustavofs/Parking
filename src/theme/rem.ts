export const REM_BASE = 16;
