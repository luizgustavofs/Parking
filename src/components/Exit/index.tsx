/* eslint-disable @next/next/no-img-element */
/* eslint-disable jsx-a11y/alt-text */
import React, { useState } from 'react';

import { useTranslation } from 'react-i18next';
import ReactLoading from 'react-loading';
import { toast } from 'react-toastify';
import { useTheme } from 'styled-components';

import {
  getVehicleHistory,
  registerVehicleExit,
  registerVehiclePayment,
} from '../../services/parking';
import Button from '../Button';
import Input from '../Input';
import { ConfirmScreen, Container, DivModal, LoadingScreen } from './styles';
import { ExitProps, ExitState } from './types';

const Exit: React.FC<ExitProps> = ({ plate, setPlate }) => {
  const { t } = useTranslation();
  const { colors } = useTheme();
  const [exitState, setExitState] = useState<ExitState>('default');
  const [residenceTime, setResidenceTime] = useState('');

  const checkCanPay = async () => {
    setExitState('loading');
    const historic = await getVehicleHistory(plate);

    const freeToPay = historic.find(element => element.paid === false);

    if (freeToPay) {
      setResidenceTime(freeToPay.time);
      setExitState('payment');
    } else {
      setExitState('default');
      const hasElementPaid = historic.find(
        element => element.left === false && element.paid === true,
      );
      if (hasElementPaid) {
        toast.success(t('PAYMENT.ALREADY.MADE'));
      } else {
        toast.error(t('PAYMENT.ALREADY.NEED-VEHICLE'));
      }
    }
  };

  const checkCanExit = async () => {
    setExitState('loading');
    const historic = await getVehicleHistory(plate);

    const freeToExit = historic.find(
      element => element.paid === true && element.left === false,
    );

    if (freeToExit) {
      setExitState('exit');
    } else {
      setExitState('default');
      const notPaid = historic.find(
        element => element.left === false && element.paid === false,
      );
      if (notPaid) {
        toast.error(t('EXIT.CHECK.ERROR'));
      } else {
        toast.error(t('EXIT.ALREADY.NEED-PAYMENT'));
      }
    }
  };

  const registerPaymentPlate = async () => {
    setExitState('loading');
    const registerPayment = await registerVehiclePayment(plate);

    if (registerPayment) {
      setExitState('success');
      toast.success(t('EXIT.PAYMENT.SUCCESS'));
      setTimeout(() => setExitState('default'), 2000);
    } else {
      toast.error(t('EXIT.PAYMENT.ERROR'));
      setExitState('default');
    }
  };

  const registerExitPlate = async () => {
    setExitState('loading');
    const registerPayment = await registerVehicleExit(plate);

    if (registerPayment) {
      setExitState('success');
      toast.success(t('EXIT.SUCCESS'));
      setTimeout(() => setExitState('default'), 2000);
    } else {
      setExitState('error');
      toast.error(t('EXIT.ERROR'));
    }
  };

  const handleHistoric = () => {
    window.location.replace(`/historic?plate=${plate}`);
  };

  return (
    <Container>
      {exitState !== 'default' && (
        <DivModal>
          <div>
            {exitState === 'loading' && (
              <LoadingScreen>
                <ReactLoading
                  type={'spokes'}
                  color={colors.primary}
                  height={130}
                  width={130}
                />
                <h4>{t('ENTRY.LOADING')}</h4>
              </LoadingScreen>
            )}
            {exitState === 'success' && (
              <ConfirmScreen>
                <img src={'/round-done-button.svg'} />
                <h4>{t('ENTRY.PAYDOUT')}</h4>
              </ConfirmScreen>
            )}
            {exitState !== 'success' && exitState !== 'loading' && (
              <>
                <h3>
                  {exitState === 'payment'
                    ? t('EXIT.MODAL.LABEL-CONFIRM')
                    : t('EXIT.MODAL.LABEL-EXIT')}
                </h3>
                <h1>{plate}</h1>
                <h3>
                  {t('EXIT.RESIDENCE-TIME')}
                  {residenceTime}
                </h3>
                <Button
                  testID="button-payment-or-exit-success"
                  primary
                  color={colors.quaternary}
                  onClick={
                    exitState === 'payment'
                      ? registerPaymentPlate
                      : registerExitPlate
                  }>
                  {exitState === 'payment'
                    ? t('EXIT.MODAL.CONFIRM')
                    : t('EXIT.MODAL.EXIT')}
                </Button>
                <Button
                  withOutBG
                  color={colors.primary}
                  onClick={() => setExitState('default')}>
                  {t('EXIT.MODAL.RETURN')}
                </Button>
              </>
            )}
          </div>
        </DivModal>
      )}

      <Input
        testID="input-plate"
        label={t('ENTRY.INPUT.LABEL')}
        placeholderText={t('ENTRY.INPUT.PLACEHOLDER')}
        value={plate}
        onChange={value => {
          setPlate(value);
          setExitState('default');
        }}
        color={exitState === 'error' ? colors.error : colors.default}
      />
      <Button
        testID="button-confirm-payment"
        primary
        disabled={plate === '' || plate.includes('_')}
        color={colors.quaternary}
        onClick={checkCanPay}>
        {t('ENTRY.PAYMENT.BUTTON')}
      </Button>
      <Button
        testID="button-confirm-left"
        secondary
        disabled={plate === '' || plate.includes('_')}
        color={colors.quaternary}
        onClick={checkCanExit}>
        {t('ENTRY.EXIT')}
      </Button>
      <Button
        testID="button-open-historic"
        withOutBG
        disabled={plate === '' || plate.includes('_')}
        color={colors.primary}
        onClick={handleHistoric}>
        {t('ENTRY.HISTORIC')}
      </Button>
    </Container>
  );
};

export default Exit;
