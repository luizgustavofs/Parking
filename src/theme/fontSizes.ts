import { REM_BASE } from './rem';

export const fontSizes = {
  h1: 38 / REM_BASE,
  h2: 24 / REM_BASE,
  h3: 16 / REM_BASE,
  h4: 15 / REM_BASE,
  h5: 12 / REM_BASE,
};
