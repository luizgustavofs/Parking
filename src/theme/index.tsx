import { ReactNode } from 'react';

import { ThemeProvider } from 'styled-components';

import { colors } from './colors';
import { fontSizes } from './fontSizes';
import { spaces } from './spaces';

export const theme = {
  colors,
  spaces,
  fontSizes,
};

const ProjectThemeProvider: React.FC<{ children: ReactNode }> = ({
  children,
  ...props
}) => {
  return (
    <ThemeProvider theme={theme} {...props}>
      {children}
    </ThemeProvider>
  );
};

export { ProjectThemeProvider };
