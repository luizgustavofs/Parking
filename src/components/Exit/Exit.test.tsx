import { fireEvent } from '@testing-library/dom';
import { render } from '@testing-library/react';
import AxiosMockAdapter from 'axios-mock-adapter';
import { act, create } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';

import Exit from '.';
import {
  MOCK_HISTORIC_COMPLETE,
  MOCK_HISTORIC_WITHOUT_LEFT,
  MOCK_HISTORIC_WITHOUT_PAYMENT,
} from '../../_mock_/mockHistoric';
import { api } from '../../api';
import {
  mockToastError,
  mockToastSuccess,
  replaceMock,
} from '../../tests/setupTests';
import { theme } from '../../theme';

describe('Exit', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with variants plate and setPlate', () => {
    const { toJSON } = create(
      <ThemeProvider theme={theme}>
        <Exit plate="AAA-0000" setPlate={jest.fn} />
      </ThemeProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should button be disable if not have plate', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Exit plate="" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm-payment');

    expect(button).toHaveProperty('disabled', true);
  });

  it('should make the payment of the plate when the button is clicked', async () => {
    var mock = new AxiosMockAdapter(api);

    mock.onGet('/parking/AAA-0000').reply(200, MOCK_HISTORIC_WITHOUT_PAYMENT);
    mock.onPost('/parking/AAA-0000/pay').reply(200, {});

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Exit plate="AAA-0000" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm-payment');
    await act(async () => {
      await fireEvent.click(button);
    });

    const buttonPay = getByTestId('button-payment-or-exit-success');
    await act(async () => {
      await fireEvent.click(buttonPay);
    });

    expect(mockToastSuccess).toHaveBeenCalled();
  });

  it('should not make the payment of the plate when the button is clicked', async () => {
    var mock = new AxiosMockAdapter(api);

    mock.onGet('/parking/AAA-0000').reply(200, MOCK_HISTORIC_COMPLETE);
    mock.onPost('/parking/AAA-0000/pay').reply(200, {});

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Exit plate="AAA-0000" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm-payment');
    await act(async () => {
      await fireEvent.click(button);
    });

    expect(mockToastError).toHaveBeenCalled();
  });

  it('should perform the left of the plate when the button is clicked', async () => {
    var mock = new AxiosMockAdapter(api);

    mock.onGet('/parking/AAA-0000').reply(200, MOCK_HISTORIC_WITHOUT_LEFT);
    mock.onPost('/parking/AAA-0000/out').reply(200, {});

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Exit plate="AAA-0000" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm-left');
    await act(async () => {
      await fireEvent.click(button);
    });

    const buttonExit = getByTestId('button-payment-or-exit-success');
    await act(async () => {
      await fireEvent.click(buttonExit);
    });

    expect(mockToastSuccess).toHaveBeenCalled();
  });

  it('should not perform the lef of the plate when the button is clicked', async () => {
    var mock = new AxiosMockAdapter(api);

    mock.onGet('/parking/AAA-0000').reply(200, MOCK_HISTORIC_COMPLETE);
    mock.onPost('/parking/AAA-0000/out').reply(200, {});

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Exit plate="AAA-0000" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm-left');
    await act(async () => {
      await fireEvent.click(button);
    });

    expect(mockToastError).toHaveBeenCalled();
  });

  it('should open the historic of the plate when the button is clicked', () => {
    var mock = new AxiosMockAdapter(api);

    mock.onGet('/parking/AAA-0000').reply(200, MOCK_HISTORIC_COMPLETE);
    mock.onPost('/parking/AAA-0000').reply(200, {});

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Exit plate="AAA-0000" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-open-historic');
    fireEvent.click(button);

    expect(replaceMock).toHaveBeenCalledWith('/historic?plate=AAA-0000');
  });

  it('should change de value input if change plate', () => {
    var mock = new AxiosMockAdapter(api);

    mock.onGet('/parking/AAA-0000').reply(200, MOCK_HISTORIC_COMPLETE);
    mock.onPost('/parking/AAA-0000').reply(200, {});

    const changePlateMock = jest.fn();

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Exit plate="AAA-0000" setPlate={changePlateMock} />
      </ThemeProvider>,
    );

    const input = getByTestId('input-plate');
    fireEvent.change(input, { target: { value: 'NEW-0000' } });

    expect(changePlateMock).toHaveBeenCalledWith('NEW-0000');
  });
});
