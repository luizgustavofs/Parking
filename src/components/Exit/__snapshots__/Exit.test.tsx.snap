// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exit should render correctly with variants plate and setPlate 1`] = `
<div
  className="sc-dkzDqf ckliYV"
>
  <div
    className="sc-gsnTZi bYaekj"
    color="#000000"
  >
    <h3>
      ENTRY.INPUT.LABEL
    </h3>
    <input
      color="#000000"
      data-testid="input-plate"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onPaste={[Function]}
      placeholder="ENTRY.INPUT.PLACEHOLDER"
      value="AAA-0000"
    />
  </div>
  <button
    className="sc-bczRLJ aoxZq"
    color="#A769B2"
    data-testid="button-confirm-payment"
    disabled={false}
    onClick={[Function]}
  >
    <h4>
      ENTRY.PAYMENT.BUTTON
    </h4>
  </button>
  <button
    className="sc-bczRLJ htJogj"
    color="#A769B2"
    data-testid="button-confirm-left"
    disabled={false}
    onClick={[Function]}
  >
    <h4>
      ENTRY.EXIT
    </h4>
  </button>
  <button
    className="sc-bczRLJ kGFRLF"
    color="#4DD0E1"
    data-testid="button-open-historic"
    disabled={false}
    onClick={[Function]}
  >
    <h4>
      ENTRY.HISTORIC
    </h4>
  </button>
</div>
`;
